docs/fuel-sync-design.md

```diff
@@ -0,0 +1,344 @@
+services inside and using the sync/async mechanism to interact.
+1. Every service knows nothing about other services. It can be interpreted 
+as: each service struct using channels for communication and knowing 
+nothing about subscribers/listeners.
+1. Every service **may** know about descending services but nothing about 
+ascending services. As an example: P2P service knows nothing about its 
+listeners(ascending services), and it only propagates information via 
+channels. But FS knows about P2P(an example of descending service)
+and may request some information directly by calling the sync/async methods. 
+The same applies to BI. FS notifies all subscribers that it synced a valid block(sealed or not sealed).
+
+Those rules are rough, and services can have cases where some functionality 
+requires direct access and some pub-sub model. So it should be decided 
+individually. But after comparison, the third rule covers all our needs 
+for now, and the description below will use this rule. We can stop 
+following this rule if it makes development harder, adds some complexity, 
+or forces us to write ugly code with workarounds.
+
+### P2P service requirements
+
+The P2P service should not be aware of FS but still should provide some 
+functionality that the FS can use. As a solution, P2P can implement the 
+`BlocksFetcher` trait.
+
+```rust
+pub enum BlockBroadcast {
+    NewSealedHeaders([Sealed<FuelBlockHeader>]),
+    NewUnsealedHeaders([FuelBlockHeader]),
+    NewSealedBlocks([Sealed<FuelBlock>]),
+    NewUnsealedBlocks([FuelBlock]),
+    // TODO: Some PoS related events
+}
+
+pub enum BlockReason {
+    InvalidHeader,
+    InvalidBody,
+    // TODO: ?
+}
+
+pub enum Verdict<Reason> {
+    Positive,
+    Negative(Reason),
+}
+
+pub enum FetchRequest {
+    HeadersByRanges([Range<BlockHeight>]),
+    HeadersByHash([Bytes32]),
+    BlocksByRanges([Range<BlockHeight>]),
+    BlocksByHash([Bytes32]),
+    /// Ask neighbors about the latest block height
+    Heartbeat,
+}
+
+// TODO: `D` should implement some bound that allows to serialize it for gossiping.
+pub trait Gossiper<D> {
+    /// Gossips some `data` to the network.
+    fn gossip(&self, data: &D);
+}
+
+pub trait Punisher<Reason> {
+    /// Anyone can report the validity or invalidity of the data. The `Punisher` decide how to process it.
+    ///
+    /// Example: If the transaction or blocker header is invalid, the punisher can decrease the reputation 
+    /// of the data provider.
+    // TODO: `D` should implement some bound that allows retrieving the data's signer and allowing to broadcast it.
+    fn report<D>(&mut self, data: D, verdict: Verdict<Reason>);
+}
+
+pub trait BlocksFetcher: Punisher<BlockReason> + Gossiper {
+    /// Returns structure that can be used to subscribe for `BlockBroadcast`.
+    fn sender(&self) -> &Sender<BlockBroadcast>;
+
+    /// Fetches the data(somehow, maybe in parallel) and broadcasts ready parts.
+    fn fetch(&mut self, request: FetchRequest);
```

Comment:
This may be a good usecase for [FuturesUnordered](https://docs.rs/futures/latest/futures/stream/futures_unordered/struct.FuturesUnordered.html), which allows joining on a set of parallel tasks.